### ISSUE TITLE
Decide on a language/framework

### ISSUE THREAD

**Issue by wschella** [ZeusWPI/HAAR#2]
Options are:

- Python/Flask
  (Feliciaan)
- Haskell/Yesod
  ()
- Java/Spring
  ()
- Ruby/Rails
  (R13N)
- Rust/(Rocket, Iron, Gotham)
  ()
- Elixir/Phoenix
  ()

To cast a vote, edit the issue, and write your name between the brackets. Add suggestions as you please.

**Comment by wschella** [comment]
Advantages and disadvantages IMO are:

- Python/Flask: 
  - Pro: 
     - Quick and dirty, will get the quickest results.
     - First graders now Python
     - Integrates nicely with resto scraper (Python)
  - Cons: 
     - Tough to maintain and expand
     - More bugprone
     - Harder to multithread
- Haskell/Yesod: 
  - Pros: 
     - Haskell is an ideal language to formally reason about data and data modeling.
     - More manageable Haskell project (<=> 12Urenloop) for new people. 
     - Very safe. Fewer weird bugs.
  - Cons: 
      - Not trivial.
 - Java/Spring: 
    - Pros:
      - Everybody (including 1st graders) knows Java
      - Our first Java (excluding Hydra Android) project
      - Good balance between easy and safe
      - Nice intro to Java for Hydra
   - Cons:
      - Java.
      - Everybody knows Java.
 - Ruby/Rails
    - Pros: 
      - Quick to develop. Great ecosystem.
      - Lot's of local experience and projects with Ruby on Rails.
    - Cons:
      - Another Ruby project
      - Same problems as Python
 -  Rust/(Rocket, Iron, Gotham)
    - Pros: 
      - Rust is da shit.
      - Safe yet fast endproduct.
      - First project with this stack. (We do have MOZAIC, with a networked rust stack though)
    - Cons:
      - Slower to develop
      - Might be overkill
      - Less mature ecosystem

**Comment by feliciaan** [comment]
My preference is for the python/flask combo, because we already have a suit of python scripts to scrape various sites.

Flask excels in small server applications like this.

**Comment by wschella** [comment]
As time goes by, this might become a medium sized project in terms of complexity tho. There are a lot of cool things we could do. Agreed that the scripts integrate best with a Python server, but they can just be called from another language as well and dump their output to a file.

**Comment by rien** [comment]
I'm also throwing my opinion in this discussion by adding and voting for Ruby. It is a very developper-friendly language and there are a lot of frameworks with excelent support & gems. If there is need a database for example, rails-api would be excellent for the job.

That aside, this should ultimately be decided by the people who are going to write the API. If @niknetniko wants to write this in PHP, I'm not stopping him.

**Comment by niknetniko** [comment]
I like Haskell as a language, however I have no experience in actually using it for something useful.

I also like Java (and Spring), but I know that a lot of people have some degree of _aversion_ to Java, so I don't want everybody to puke every time they open the repo 😢. It might also be overkill for an application of this size.

Python and Ruby are OK.

**Comment by Procrat** [comment]
I don't want to express my opinion here, but you could consider seeing this as an opportunity to try out something new, like [Rocket](https://rocket.rs/), [Iron](http://ironframework.io/), [Gotham](https://gotham.rs/) or [Phoenix](http://phoenixframework.org/).

**Comment by wschella** [comment]
Suggestions have been added. Personally I favor a strongly typed language, as we will be doing mostly data aggregation and manipulation, which benefits strongly from formal reasoning. As I'm doing Rust already, my personal choice would be Haskell, but I feel Java will benefit the Hydra project more by being a gateway for new/external people.

**Comment by Procrat** [comment]
Oh, no, it wasn't my intention to propose those frameworks specifically, but rather to suggest that this could be an opportunity to go wild and try out something new, whatever that may be. :smile:

**Comment by feliciaan** [comment]
Maybe something like perfect.org <http://perfect.org/> because it's called perfect :p

**Comment by rien** [comment]
I agree with @Procrat, this is ideal to get to know a new framework/programming language. For example: [crystal](https://crystal-lang.org/) (ruby-based, compiled language) with [kemal](https://github.com/kemalcr/kemal) is something I've wanted to try out for a while now.

It has the style and philosophy of ruby and the advantages of a compiled language. A disadvantage however, is the fact that it's still a very young language. So support and the available plugins may not be optimal.

**Comment by wschella** [comment]
I'm a big fan of using [Reason](https://reasonml.github.io/) combined with the Node ecosystem.
